dotnet/src/Extensions/Planning.FlowPlanner/Model/FlowStep.cs

```diff
@@ -0,0 +1,155 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+#pragma warning disable IDE0130 // Namespace does not match folder structure
+// ReSharper disable once CheckNamespace
+namespace Microsoft.SemanticKernel.Planning.Flow;
+#pragma warning restore IDE0130 // Namespace does not match folder structure
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+
+public class FlowStep
+{
+    private List<string> _requires = new();
+
+    private List<string> _provides = new();
+
+    private Dictionary<string, Type?> _skillTypes = new();
+
+    private Func<IKernel, Dictionary<object, string?>, IEnumerable<object>>? _skillsFactory;
+
+    public FlowStep(string goal, Func<IKernel, Dictionary<object, string?>, IEnumerable<object>>? skillsFactory = null)
+    {
+        this.Goal = goal;
+        this._skillsFactory = skillsFactory;
+    }
+
+    /// <summary>
+    /// Deserialization only
+    /// </summary>
+    public FlowStep() : this(string.Empty)
+    {
+    }
+
+    public string Goal { get; set; }
+
+    public List<string> Requires
```

Comment:
Thanks for the PR, will take some time to go through it. I'd like to understand these pieces better (docs will help) and whether it would make sense to include as part of `Plan` or `SKFunction` objects.

Reply:
Thanks Lee. Just completed the docs for the data models.
I did explore `Plan` in the beginning. however there are some gaps so I am proposing a different `IPlan` implementation. // haven't really implemented the marker interface yet.

Here are the quoted comments from `Flow.cs`. 
```
/// Principles:
/// 1. The model should be decoupled from execution status
/// 2. The model is mutable to allow dynamic changes
/// 3. The model doesn't enforce any execution order as long as the dependencies are satisfied.
```

Besides these, there are some other differences
1. for each step in the flow, there could be a set of skills to use to achieve the step goal.
2. each step should support multi-in, multi-out
3. execution context for each step should be highly isolated to avoid noise and waste of tokens